### ISSUE TITLE
Triggered based on some but not all inputs

### ISSUE THREAD

**Issue by dwhitena** [pachyderm/pachyderm#2296]
Let's say that we have some reference data that is being generated in a first pipeline (maybe persisted models), and that reference data is to be used in one or more other pipelines.  The follow potential feature was brought up in users Slack:

It would be nice to trigger pipelines automatically on some of the inputs but not on other of the inputs.  For example, you may want to utilize the latest persisted model when you run a Pipeline on new events coming into a first repo.  But you might not want to run that pipeline each time the persisted model is updated.

Right now it appears that users are working around this by egressing or exporting reference data to a globally accessible object store bucket (outside of Pachyderm), then re-downloading the data when it is needed.

**Comment by JoeyZwicker** [comment]
So this would be a `non-triggering` flag that they'd specify as part of the `input` spec in the pipeline?

In the scenario you described, what WOULD they want to happen when the persisted model (PM) is updated? 

Do new updates to there other input use the old persisted model? In other words, do they want to have one input repo be "fixed" to a specific commit that is manually updated in the pipeline spec when they want to change it?

Or does the other input repo always use the "head" persisted model, but just doesn't trigger based on PM updates? 

What if there are multiple new versions of the PM? Run on all or just HEAD?

**Comment by dwhitena** [comment]
Yes, a repo-level `non-triggering` flag of some sort.

In this scenario they wouldn't want anything to happen when the persisted model is updated.  They would just want the latest model to be utilize for new commits on the other input, whenever those new commits happen.

This is related to the discussions we have had about `update-pipeline` and models.  Sometimes when you update a model, you don't want to re-calculate results for previous input.  You just want it to be utilized for new inputs.  This is basically the parallel of that when, instead of changing your code, you change your training data, for example.

If there are multiple new versions of the PM, you would just run on the HEAD.  The idea is to always use the latest PM when you make predictions, but not update for old input whenever the PM changes.

**Comment by kareemk** [comment]
Confirming that @dwhitena is describing our usecase accurately. Thank you.

**Comment by kareemk** [comment]
Reading through this raises another issue, is it possible for a job to then write to this repo (as well as its default repo)?

**Comment by JoeyZwicker** [comment]
@kareemk can you clarify your question? In general, a job/pipeline can only write out to a single repo.

**Comment by pragmasoft-ua** [comment]
We're considering integrating Pachyderm as a pipeline engine into our product, and we also need better control over what can trigger a pipeline. Ideally would be to have a possibility to disable automatic triggering at all, for example, if customer account does not have enough balance when customer uploads new data to the repo, and also possibility to trigger pipeline manually by some api call and command. It would then work on files committed since last time it was triggered.

**Comment by JoeyZwicker** [comment]
> Ideally would be to have a possibility to disable automatic triggering at all, for example, if customer account does not have enough balance when customer uploads new data to the repo, and also possibility to trigger pipeline manually by some api call and command. It would then work on files committed since last time it was triggered.

We're discussing adding options like this. It's not ideal yet, but `stop-pipeline` and `start-pipeline` should give you a way to implement some of this already. `stop-pipeline` basically just pauses a pipeline so it not longer triggers on commits, and then it'll catch up once you start it again. That catch up will include one job per commit, but as always, pachyderm will only process each new piece of data once even if it shows up in multiple commits.

**Comment by itssimon** [comment]
Has any progress been made on this suggestion? Such an option would be perfectly suited for our use cases as well.

**Comment by jdoliner** [comment]
Hi @itssimon unfortunately we haven't gotten this implemented, could you tell us a bit more about your use case so we can understand the requirements for this feature better and also recommend potential workarounds for you?

**Comment by itssimon** [comment]
Yeah, there probably is an obvious workaround for this, but I could use some guidance on how to best implement it. I've got a pipeline that pulls data from S3 and processes it continually. Only once a month I want to generate an aggregate report based on all the data of the previous month. If I could use a cron input crossed with a non-trigger atom input of the continually updated data that would be easy.

**Comment by jdoliner** [comment]
@itssimon @gabrielgrant  the recommended way of doing this is to have to branches in place of the "non-trigger" input. One that's actually hooked up to the pipeline that you commit to only when you want to run the pipeline (`master`) and one that's not hooked up to anything and you can commit to it whenever you want (`staging`).
Now when you want to run the pipeline with the latest version of the data available in staging you just need to copy staging to master with `pachctl copy-file repo staging / repo master /`. Copying is just rewriting pointers so it's a very efficient operation.

**Comment by itssimon** [comment]
Thanks, that should do just fine!

**Comment by JoeyZwicker** [comment]
Updating this from in-person conversations.

In cases where you have two separate inputs where you want one to be triggering and non-triggering (in functionality terms), this can be accomplished in the current system using differred processing as follows:

In this scenario, I’m assuming that data added to `A` you want to be non-triggering and data added to `B` to be triggering.

First, instead of having A and B as separate repos, you’ll need to make those directories in the same repo. Pipelines can take multiple inputs from the same repo, so that’s fine, and you’ll just use different glob patterns. So your inputs will be `repo:foo` `glob: A/` or `glob: A/*` or whatever glob patterns you’re using and `repo:foo` `glob: B/` etc.

The other piece to this is to use the deferred processing method outlined in this document: http://docs.pachyderm.io/en/latest/cookbook/deferred_processing.html.

Your pipeline inputs will both be taking the `master` branch of that repo. But when you commit data (e.g. put-file …), you should commit it to the `staging` branch (you can name that branch something other than “staging” of course). Whenever you want to trigger a job, you simply call `pachctl set-branch staging master` which will trigger a job using the HEAD commit on staging. In this case, this means that any time you commit data to `B` you call `set-branch` after that, but dont call `set-branch` when committing to `A`.

Currently, the input sources need to be in the same repo for this to work nicely. We understand that there are logical scenarios where having `A` and `B` in the same repo is annoying because there may be other pipelines that consume A and B and they prefer to be separated. Once #3267 is added, this will work as described above in separate repos as well. You'd atomic update the branches in both repos. 

One unresolved annoyance of this deferred processing method is that it requires that whoever is adding data to the repo to be aware of pipelines that are consuming that repo and follow this differed processing method. This can be very cumbersome to automate and is restrictive in a large organization where adding data to pfs and pipeline creators are very separated. This is a possible argument for still considering "passive" (e.g. non-triggering) inputs -- that way this behavior is captured in the pipeline only independent of data getting added to the repo.

**Comment by jdoliner** [comment]
This workflow is now well supported through a combination of [deferred processing](https://docs.pachyderm.io/en/latest/cookbook/deferred_processing.html) and transactions.
